client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx

```diff
@@ -45,17 +56,42 @@ interface Props {
     onDidError: (error: ErrorLike) => void
 }
 
+// The data stored in user_external_account has this structure: "Values"["<attribute names for email or username>"] > "Values"[0] > "Value": string
+const emailAttrs = new Set(['http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress', 'emailaddress', 'http://schemas.xmlsoap.org/claims/EmailAddress'])
+const usernameAttrs = new Set(['nickname', 'login', 'username', 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name'])
+
+export function getUsernameOrEmail(data: SamlExternalData): string {
+    let userNameOrEmail = ''
+    if (data.Values) {
+        const entries = Object.entries(data.Values)
+        for (const [name, att] of entries) {
+            const attribute = att as Attribute
+            if (usernameAttrs.has(name)) {
+                userNameOrEmail = attribute.Values.length > 0 && attribute.Values[0].Value || ''
+                break
+            }
+            if (emailAttrs.has(name)) {
+                userNameOrEmail = attribute.Values.find((val: AttributeValue) =>
+                    val.Value.includes('@')
+                )?.Value || ''
+            }
+        }
+    }
+
+    return userNameOrEmail
+}
+
 const getNormalizedAccount = (
     accounts: Partial<Record<string, UserExternalAccount>>,
     authProvider: AuthProvider
 ): NormalizedMinAccount => {
     // kind and type match except for the casing
-    const kind = authProvider.serviceType.toLocaleUpperCase() as ExternalServiceKind
+    const kind = authProvider.serviceType as ExternalAccountKind
```

Comment:
It might be better to use a type guard here and make sure that `.serviceType` doesn't contain any unwanted fields. Something like 
```suggestion
    ): NormalizedAccount | null => {
    if (authProvider.serviceType === 'builtin' || authProvider.serviceType === 'http-header') {
        return null
    }
```

And then

Reply:
Wondering if there's a comment missing here...